source/grid/manifold_lib.cc

```diff
@@ -421,20 +421,46 @@ get_new_point (const ArrayView<const Point<spacedim>> &vertices,
 
     // If the candidate happens to coincide with a normalized
     // direction, we return it. Otherwise, the Hessian would be singular.
-    boost::container::small_vector<Tensor<1, 3>, 100> directions(n_points);
+    boost::container::small_vector<Tensor<1, 3>, 100> directions;
+    boost::container::small_vector<double, 100> merged_weights;
     for (unsigned int i=0; i<n_points; ++i)
       {
-        for (unsigned int c = 0; c < spacedim; ++c)
-          directions[i][c] = vertices[i][c] - center[c];
-        const double norm = directions[i].norm();
+        Tensor<1,spacedim> direction(vertices[i]-center);
+        const double norm = direction.norm();
         Assert(norm != 0.,
                ExcMessage("One of the vertices coincides with the center. "
                           "This is not allowed!"));
-        directions[i] /= norm;
-        if ((xVec - directions[i]).norm_square() < tolerance*tolerance)
+        direction /= norm;
+        if ((candidate - direction).norm_square() < tolerance*tolerance)
           return center + rho * candidate;
+
+        // append direction. check if the normalized candidate direction is
+        // the same as a previous direction (to a tighter tolerance (1e-14)^2
+        // than the outer ones to really not miss anything) -> in that case we
+        // can simply add the weights. Since the trigonometric functions used
+        // below are quite expensive, it makes sense to merge the points here,
+        // even if this search loop is of quadratic complexity loop (but we
+        // rarely have more than 9 points)
+        Tensor<1,3> direction_3d;
```

Comment:
Actually, we know at this point that `spacedim==3`. Unfortunately, the compiler doesn't and would not let us call `compute_normal`,  `projected_direction` and `apply_exponential_map`... I am not sure if it is worth to find a workaround to avoid the copy.

Reply:
The copy is not expensive I guess, at least compared to all the other things we are doing here. So I would keep things as they are.